### PR TITLE
Support package with SemVer2 version with buildMetadata

### PR DESCRIPTION
**Pull request by JoC0de**
Fixes: #540

### CODE REVIEW

File: src/NuGetForUnity/Editor/NugetPackageIdentifier.cs [hunk 1]
```diff
@@ -43,30 +45,85 @@ public NugetPackageIdentifier(string id, string version)
...
+                    normalizedVersion = string.Empty;
+                    FullVersion = string.Empty;
+                    IsPrerelease = false;
+                    IsMinInclusive = false;
+                    IsMaxInclusive = false;
+                    HasVersionRange = false;
+                    return;
+                }
+
+                if (normalizedVersion == value)
```
Comment (JoC0de) [inline src/NuGetForUnity/Editor/NugetPackageIdentifier.cs:70]:
```suggestion
                if (FullVersion == value)
```

File: src/NuGetForUnity/Editor/NugetPackageIdentifier.cs [hunk 2]
```diff
@@ -89,7 +151,12 @@ public int CompareTo(NugetPackageIdentifier other)
                 return string.Compare(Id, other.Id);
             }
 
-            return CompareVersion(other.Version);
+            if (HasVersionRange)
```
Comment (JoC0de) [inline src/NuGetForUnity/Editor/NugetPackageIdentifier.cs:154]:
```suggestion
            if (HasVersionRange || other.HasVersionRange)
```

File: src/NuGetForUnity/Editor/NugetPackageIdentifier.cs [hunk 3]
```diff
@@ -345,127 +403,202 @@ public int CompareVersion(string otherVersion)
...
+
+            private readonly string buildMetadata;
+
+            public string PreRelease { get; }
+
+            /// <summary>
+            ///     Initializes a new instance of the <see cref="SemVer2Version" /> struct.
+            /// </summary>
+            /// <param name="dummy">Dummy to allow calling this constructor.</param>
+            public SemVer2Version(bool dummy)
```
Comment (igor84) [inline src/NuGetForUnity/Editor/NugetPackageIdentifier.cs:435]: Maybe a better option would be a `static SemVer2Version Create()` method?

File: src/NuGetForUnity/Editor/NugetPackageIdentifier.cs [hunk 4]
```diff
@@ -345,127 +403,202 @@ public int CompareVersion(string otherVersion)
...
-            catch (Exception)
+
+            /// <summary>
+            ///     Returns the version number as a string in the form <see cref="major" />.<see cref="minor" />.<see cref="patch" />-<see cref="PreRelease" />+
+            ///     <see cref="buildMetadata" />.
+            ///     The <see cref="buildMetadata" /> can be removed.
+            /// </summary>
+            /// <param name="withBuildMetadata">If <c>true</c> the <see cref="buildMetadata" /> is included.</param>
+            /// <returns>The formatted string.</returns>
+            public string ToString(bool withBuildMetadata = false)
```
[the remark this answers is not in the document]
Comment (JoC0de) [reply src/NuGetForUnity/Editor/NugetPackageIdentifier.cs:571]: No you can remove the `= false`. I need to install ReSharper 😄

File: src/NuGetForUnity/Editor/NugetPackageIdentifier.cs (hunk 3 above)
Comment (JoC0de) [reply src/NuGetForUnity/Editor/NugetPackageIdentifier.cs:435]: I never used `struct` in a old C# language version where there Default constructor is not supported.
I just didn't hat that in mind. If you like you can change it.

**Pull request merged**